### DISCUSSION TITLE
How to describe a fabric-based 16550 UART to the linux device tree.

### DISCUSSION THREAD

**Discussion by JV73** [polarfire-soc/polarfire-soc-documentation#197] [Linux build systems for PolarFire SoC] [answered]
I've got an 16550 UART implemented in the PolarFireSoC FPGA fabric and connected to the MSS via an APB interface (FIC3).  I've got my UART defined in the device tree, and aliased to serial5 as follows.  It shows up in linux as ttyS0.

	uart5: serial@41000C00 {
		compatible = "ns16550a";
		reg = <0x0 0x41000C00 0x0 0x20>;
		reg-io-width = <4>;
		reg-shift = <2>;
		interrupt-parent = <&plic>;
		interrupts = <128>;
		clocks = <&clkcfg CLK_FIC3>;
		current-speed = <115200>;
		status = "okay";
	};

I can connect to it with 'screen'.  I'm setting the baud (on the screen command line) to 115200, and screen is writing x00A3 to into the DIV registers in the fabric UART, which gives a 768 Hz bit rate.  To actually get 115200 Baud, it needs to write 0x0001. In the dts description (above), I tell it: clocks = <&clkcfg CLK_FIC3>;  but it seems now like I need to tell it that the 'base clock rate' is 115200 Hz - or more likely, x16 = 1843200  or maybe 3686400 depending on the drivers assumptions about the 'clock' to baud ratio.

On a similar note: is there a way to specify an rx buffer size for the kernel driver ?  The external application that will connect to can burst data in blocks of 2-3k bytes so the driver may need a 4k buffer to catch all the data in case the application running under linux does not keep up with things.

Anyone know the correct syntax that will satisfy Yocto and the built-in "ns16550a" driver ?

**Comment by JV73** [comment, marked as answer]
After a bit of trial and error, it turns out that replacing "clocks = <&clkcfg CLK_FIC3>;" with "clock-frequency = <1843200>;" results in the correct 'base baud rate' of 115200 baud.  This is necessary because the particular core implements multiple serial ports by time-slicing the FIC3 clock across the number of channels implemented. So the 'base baud rate' (the speed when DIV = 1) depends on both the FIC3 rate AND the number of ports implemented in hardware.  So in this case,, the base baud rate is 115200 and I need to tell the driver that the clock is 16x115200 = 1843200 Hz.

Also - the default ICICLE-Kit-AMP yocto build defaults to a limit of 4 serial ports.  My implementation can include up to 32 UARTs in the fabric + up to 4 of the five MSS_UARTs for a total of 36.  Since all of the UARTs use the ns16550a driver, the maximum number of ports that linux will probe for during boot is set in the 'linux/drivers/tty/serial/8250/kconfig' file.  Lines 183 -199.  There are two symbols - both need to be updated from the 'default' value of 4 to 36 in this case. 

After re-building and uploading linux with these changes, I can ssh into the ICICLE kit, see the 7 UARTs that are currently defined in the Device Tree, and use 'screen /dev/ttySx <baud>,-xon,-xoff' to connect to any of them and send/receive characters between the 'screen' session and a Terminal connected to the RX/TX signal via a USB/serial adapter.

**Reply by ConchuOD** [reply]
Hmm, sounds like we need a specific compatible for the fabric UART core to me & maybe some clock handling code on the driver side.
The clock listed in the DTS should be the same frequency as the one connected to the core, so while this may work for you it is not the "correct" way to solve the issue - possibly an entire dedicated driver is required to satisfy the clock configuration needs.

(I would not really consider this "answered" more like "worked around")
